pkg/kubelet/images/pullmanager/image_pull_manager.go

```
@@ -0,0 +1,472 @@
+	}
+
+	if !pulledRecordChanged {
+		return nil
+	}
+
+	return f.recordsAccessor.WriteImagePulledRecord(pulledRecord)
+}
+
+func (f *PullManager) RecordImagePullFailed(image string) {
+	f.decrementImagePullIntent(image)
+}
+
+// decrementImagePullIntent decreses the number of how many times image pull
+// intent for a given `image` was requested, and removes the ImagePullIntent file
+// if the reference counter for the image reaches zero.
+func (f *PullManager) decrementImagePullIntent(image string) {
+	f.intentAccessors.Lock(image)
+	defer f.intentAccessors.Unlock(image)
+
+	f.intentCounters[image]--
+	if f.intentCounters[image] > 0 {
+		return
+	}
+	delete(f.intentCounters, image)
+
+	if err := f.recordsAccessor.DeleteImagePullIntent(image); err != nil {
+		klog.ErrorS(err, "failed to remove image pull intent", "image", image)
+	}
+}
+
+func (f *PullManager) MustAttemptImagePull(image, imageRef string, podSecrets []kubeletconfiginternal.ImagePullSecret) bool {
+	if len(imageRef) == 0 {
+		return true
+	}
+
+	var imagePulledByKubelet bool
+	var pulledRecord *kubeletconfiginternal.ImagePulledRecord
+
+	err := func() error {
+		// don't allow changes to the files we're using for our decision
+		f.pulledAccessors.Lock(imageRef)
+		defer f.pulledAccessors.Unlock(imageRef)
+		f.intentAccessors.Lock(image)
+		defer f.intentAccessors.Unlock(image)
+
+		var err error
+		var exists bool
+		pulledRecord, exists, err = f.recordsAccessor.GetImagePulledRecord(imageRef)
+		switch {
+		case err == nil && pulledRecord == nil:
+			if exists, err := f.recordsAccessor.ImagePullIntentExists(image); err != nil {
+				return fmt.Errorf("failed to check existence of an image pull intent: %w", err)
+			} else if exists {
+				imagePulledByKubelet = true
+			}
+		case err == nil && pulledRecord != nil:
+			imagePulledByKubelet = true
+		case err != nil && exists: // record exists but has invalid format
```

looking at this again after some time:
the error is not propagated to the below code, we don't flip `imagePulledByKubelet` to true... Is this correct?